### PR TITLE
Migrate ReasoningIT to BDD

### PR DESCRIPTION
**Pull request by alexjpwalker**
## What is the goal of this PR?

To identify and de-duplicate integration tests that test the fundamentals of Reasoner, categorise them appropriately and migrate them to BDD scenarios.

This PR will focus exclusively on ReasoningIT.

This PR forms part of the work towards https://github.com/graknlabs/grakn/issues/5721

## What are the changes implemented in this PR?

Migrate tests away from IT files as they get moved to BDD in https://github.com/graknlabs/verification/pull/81

### CODE REVIEW

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 1]
```diff
@@ -56,277 +51,6 @@
     private static String resourcePath = "test/integration/graql/reasoner/stubs/";
 
     //The tests validate the correctness of the rule reasoning implementation w.r.t. the intended semantics of rules.
-    //The ignored tests reveal some bugs in the reasoning algorithm, as they don't return the expected results,
-    //as specified in the respective comments below.
-
-    @Test
-    public void whenMaterialising_duplicatesAreNotCreated() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:63]: migrated to negation and relation-inference

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 2]
```diff
@@ -56,277 +51,6 @@
...
-                                "$rel has inferredAttribute 'inferredRelation';" +
-                                "get;")
-                        .asGet());
-                assertEquals(25, answers.size());
-            }
-        }
-    }
-
-    @Test //Expected result: Both queries should return a non-empty result, with $x/$y mapped to a unique entity.
-    public void unificationOfReflexiveRelations() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:78]: migrated to relation-inference and match

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 3]
```diff
@@ -56,277 +51,6 @@
...
-
-                String queryString = "match (role1:$x, role2:$y) isa relation3; get;";
-                List<ConceptMap> answers = tx.execute(Graql.parse(queryString).asGet());
-                assertEquals(2, answers.size());
-            }
-        }
-    }
-
-    @Test
-    public void resourcesAsRolePlayers() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:194]: migrated to relation-inference and value-predicate

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 4]
```diff
@@ -56,277 +51,6 @@
...
-                assertEquals(1, answers4.size());
-                assertEquals(answers.size() + answers2.size() + answers3.size() + answers4.size(), answers5.size());
-                assertEquals(answers5.size() - answers4.size(), answers6.size());
-            }
-        }
-
-    }
-
-    @Test
-    public void resourcesAsRolePlayers_vpPropagationTest() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:234]: migrated to relation-inference and value-predicate

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 5]
```diff
@@ -56,277 +51,6 @@
...
-                assertEquals(3, answers3.size());
-                assertEquals(3, answers4.size());
-                assertEquals(answers5.size(), answers.size() + answers2.size() + answers3.size() + answers4.size());
-                assertEquals(answers6.size(), answers5.size() - answers4.size());
-            }
-        }
-    }
-
-    @Test //Expected result: Returns db and inferred relations + their inverses and relations with self for all entities
-    public void reasoningWithRepeatingRoles() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:272]: extremely similar to multiple existing tests

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 6]
```diff
@@ -56,277 +51,6 @@
...
-                    List<ConceptMap> answers = tx.execute(Graql.parse(queryString).asGet());
-                    assertEquals(6, answers.size());
-                    assertCollectionsNonTriviallyEqual(oldAnswers, answers);
-                }
-            }
-        }
-    }
-
-    @Test //Expected result: Relations between all entity instances including relation between each instance and itself
-    public void reasoningWithEntityTypes() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:302]: duplicate test exists in concept-inequality

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 7]
```diff
@@ -56,277 +51,6 @@
...
-
-                String uniquePairQuery = "match (role1:$x1, role2:$x2) isa uniquePairRelation; get;";
-                List<ConceptMap> uniquePairs = tx.execute(Graql.parse(uniquePairQuery).asGet());
-                assertEquals(6, uniquePairs.size());
-            }
-        }
-    }
-
-    @Test //Expected result: Timeline is correctly recognised via applying resource comparisons in the rule body
-    public void reasoningWithResourceValueComparison() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:319]: migrated to value-predicate

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 8]
```diff
@@ -358,203 +82,38 @@ public void reasoningWithRelations() {
         }
     }
 
-    @Test //Expected result: number of answers equal to specified limit (no duplicates produced)
-    public void whenReasoningWithRelationConjunctions_duplicatesNotProducesAndTypesInferredCorrectly() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:362]: migrated the useful parts to relation-inference and match

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 9]
```diff
@@ -358,203 +82,38 @@ public void reasoningWithRelations() {
...
-
-                List<ConceptMap> answers = tx.execute(Graql.parse(queryString).asGet());
-                assertEquals(6, answers.size());
-                answers.forEach(ans -> assertEquals(4, ans.size()));
-            }
-        }
-    }
-
-    @Test
-    public void whenAppendingRolePlayers_queryIsRewrittenCorrectly() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:475]: migrated to roleplayer-attachment and negation

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 10]
```diff
@@ -358,203 +82,38 @@ public void reasoningWithRelations() {
...
                         "get;"));
-                assertEquals(2, answers5.size());
 
                 assertEquals("New relations were created!", persistedRelations, tx.execute(Graql.parse("match $r isa baseRelation; get;").asGet(), false));
             }
         }
     }
 
-    @Test
-    public void whenAppendingRolePlayers_DuplicatesAreCreated() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:512]: migrated to roleplayer-attachment

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 11]
```diff
@@ -358,203 +82,38 @@ public void reasoningWithRelations() {
...
-                            .filter(thing -> thing.equals(player))
-                            .collect(Collectors.toList());
-                    assertTrue(identicalRolePlayers.size() == 2);
-                });
-            }
-        }
-    }
-
-    @Test
-    public void whenCopyingRolePlayer_DuplicateRoleRetrievedCorrectly() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:532]: migrated to roleplayer-attachment

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 12]
```diff
@@ -577,129 +136,4 @@ public void whenAppendingRolePlayers_noNewRelationsAreCreated() {
...
-                assertEquals(1, answers3.size());
-                answers2.stream()
-                        .map(a -> a.project(Sets.newHashSet(new Variable("x"), new Variable("y"))))
-                        .forEach(a -> assertTrue(answers3.contains(a)));
-            }
-        }
-    }
-
-    @Test
-    public void ternaryRelationsRequiringDifferentMultiunifiers() {
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:643]: already exists in variable-role

File: test/integration/graql/reasoner/ReasoningIT.java [hunk 13]
```diff
@@ -20,8 +20,6 @@
 import com.google.common.collect.Iterables;
```
Comment (alexjpwalker) [inline test/integration/graql/reasoner/ReasoningIT.java:20]: Comments have been added to the tests so the reviewer is able to figure out where they were migrated to.
If unspecified, they were migrated to `relation-inference.feature` in verification.

File: test/behaviour/resolution/ResolutionSteps.java [hunk 14]
```diff
@@ -95,13 +96,28 @@ public void reasoned_keyspace_answer_size_is(final int expectedCount) {
         }
     }
 
+    @Then("answers are consistent across {int} executions in reasoned keyspace")
+    public void reasoned_keyspace_answers_are_consistent_across_n_executions(final int executionCount) {
```
Comment (flyingsilverfin) [inline test/behaviour/resolution/ResolutionSteps.java:100]: why are we prefixing with "reasoned_keyspace"? is this our current style?

File: test/integration/graql/reasoner/ReasoningIT.java (hunk 12 above)
Comment (flyingsilverfin) [reply test/integration/graql/reasoner/ReasoningIT.java:643]: lets go over this one  - there's a reason these are all grouped together!

Comment (alexjpwalker) [reply test/integration/graql/reasoner/ReasoningIT.java:643]: They are also grouped together in variable-roles.feature, in the scenario:
`Scenario: variable roles are correctly mapped to answers for a ternary relation with 3 possible roleplayers`

File: test/behaviour/resolution/ResolutionSteps.java (hunk 14 above)
Comment (alexjpwalker) [reply test/behaviour/resolution/ResolutionSteps.java]: Updated to answers_are_..._in_reasoned_keyspace

**Pull request merged**